### PR TITLE
Install distributor now normalizes uid, gid, and filesystem permissions.

### PR DESCRIPTION
**Pull request by mhrivnak**
Also resurrected a bunch of unit tests for the install distributor that had
been removed during the mongoengine conversion.

fixes #1289

https://pulp.plan.io/issues/1289

### CODE REVIEW

File: pulp_puppet_plugins/pulp_puppet/plugins/distributors/installdistributor.py [hunk 1]
```diff
@@ -147,11 +180,7 @@ def publish_repo(self, repo, publish_conduit, config):
                 _('failed to move temporary destination to destination directory: %s') % str(e),
                 self.detail_report.report)
 
-        # return some kind of report
-        if self.detail_report.has_errors:
-            return publish_conduit.build_failure_report(_('failed'), self.detail_report.report)
```
Comment (mhrivnak) [inline pulp_puppet_plugins/pulp_puppet/plugins/distributors/installdistributor.py]: This was an impossible, and thus untestable, circumstance.

File: pulp_puppet_plugins/pulp_puppet/plugins/distributors/installdistributor.py [hunk 2]
```diff
@@ -28,6 +27,40 @@ def entry_point():
...
+        This also adds a basic level of filesystem permissions to each file and directory extracted,
+        to ensure that they can be accessed and moved around as required.
+
+        :return:    a new TarInfo instance representing a file in an archive
+        :rtype:     tarfile.TarInfo
+        """
+        ret = super(NormalizingTarInfo, cls).fromtarfile(*args, **kwargs)
+        # this causes the uid and gid to be set matching the process doing the extraction
+        ret.uid = 0
+        ret.gid = 0
```
Comment (rbarlow) [inline pulp_puppet_plugins/pulp_puppet/plugins/distributors/installdistributor.py]: Are not root privileges required for any process to change uid and gid ownership of files? It seems that it is unnecessary for Pulp to try to change these values, unless someone is planning to run Pulp as root (in which case this code will cause the files to be owned by root anyway).

File: pulp_puppet_plugins/pulp_puppet/plugins/distributors/installdistributor.py [hunk 3]
```diff
@@ -28,6 +27,40 @@ def entry_point():
...
+    Specifically, it is important to guarantee that puppet can read the extracted files.
+    """
+    min_file_perms = int('0644', base=8)
+    min_dir_perms = int('0755', base=8)
+
+    @classmethod
+    def fromtarfile(cls, *args, **kwargs):
+        """
+        This removes the uid and gid from the returned TarInfo file, causing those values to be set
+        based on the process currently running.
```
Comment (rbarlow) [inline pulp_puppet_plugins/pulp_puppet/plugins/distributors/installdistributor.py]: I'd recommend removing this sentence too.

File: pulp_puppet_plugins/pulp_puppet/plugins/distributors/installdistributor.py (hunk 1 above)
Comment (rbarlow) [inline pulp_puppet_plugins/pulp_puppet/plugins/distributors/installdistributor.py]: Can you elaborate on why this is impossible?

File: pulp_puppet_plugins/pulp_puppet/plugins/distributors/installdistributor.py [hunk 4]
```diff
@@ -28,6 +27,40 @@ def entry_point():
...
+        :rtype:     tarfile.TarInfo
+        """
+        ret = super(NormalizingTarInfo, cls).fromtarfile(*args, **kwargs)
+        # this causes the uid and gid to be set matching the process doing the extraction
+        ret.uid = 0
+        ret.gid = 0
+        if ret.isfile():
+            ret.mode |= cls.min_file_perms
+        elif ret.isdir():
+            ret.mode |= cls.min_dir_perms
```
Comment (rbarlow) [inline pulp_puppet_plugins/pulp_puppet/plugins/distributors/installdistributor.py]: It's always fun to do bitwise operations!

**Comment by rbarlow** [comment]
I recommend splitting this PR into two PRs, since part of it fixes a targeted issue (the permission issue) and the other part is introducing unrelated unit tests. I nearly lost my mind last week due to needing to deal with a past commit in pulp_docker that was not atomic ☺. LGTM.

File: pulp_puppet_plugins/pulp_puppet/plugins/distributors/installdistributor.py (hunk 1 above)
Comment (mhrivnak) [inline pulp_puppet_plugins/pulp_puppet/plugins/distributors/installdistributor.py]: because if line 151 was True, line 131 would also have been True, and would have returned.

File: pulp_puppet_plugins/pulp_puppet/plugins/distributors/installdistributor.py (hunk 2 above)
Comment (mhrivnak) [inline pulp_puppet_plugins/pulp_puppet/plugins/distributors/installdistributor.py]: If the worker process is run by root, the default behavior is to attempt to write files with the uid and gid from the tarball. I think that is worth guarding against, and this is a very simple behavior that is not likely to cause confusion.

The docs also have a vague implication that other platforms may allow "normal" users to assign a different uid or gid. There are no additional details, and this is admittedly unlikely even for an edge case, but we can guard against that also with these very simple 2 lines of code.

**Comment by mhrivnak** [comment]
The additional tests are actually very related. They are tests for the area whose behavior I was changing, albeit with a change to only one line. The area I was changing had very low test coverage, so I just resurrected and fixed the tests that got removed recently in the mongoengine conversion.

I could submit a PR resurrecting those tests, then submit a second PR fixing the bug and changing those tests to reflect the changed behavior. But given that it's such a targeted area, I don't think that's worth the time and effort.
